File: lua/utils/format_file_size.lua
local function format_file_size(bytes, precision, nbsp)
  precision = precision or 2
  if bytes == 0 then return 'n/a' end
  local units = { 'B', 'KB', 'MB', 'GB', 'TB', 'PB' }
  local pow = math.floor(math.log(bytes) / math.log(1024))
  if pow > #units - 1 then pow = #units - 1 end
  local value = bytes / (1024 ^ pow)
  local space = (nbsp == false) and ' ' or '\u00A0'
  return string.format('%.' .. precision .. 'f%s%s', value, space, units[pow+1])
end

if pcall(debug.getlocal, 4, 1) == false then
  local b = tonumber(arg and arg[1] or '0') or 0
  local p = tonumber(arg and arg[2] or '2') or 2
  print(format_file_size(b, p, true))
end

return { format_file_size = format_file_size }

File: go/utils/format_file_size.go
package main

import (
    "fmt"
    "math"
    "os"
    "strconv"
)

func formatFileSize(bytes float64, precision int, nbsp bool) string {
    if bytes == 0 {
        return "n/a"
    }
    units := []string{"B", "KB", "MB", "GB", "TB", "PB"}
    pow := int(math.Floor(math.Log(bytes) / math.Log(1024)))
    if pow > len(units)-1 {
        pow = len(units) - 1
    }
    value := bytes / math.Pow(1024, float64(pow))
    space := " "
    if nbsp {
        space = "\u00A0"
    }
    format := fmt.Sprintf("%%.%df", precision)
    return fmt.Sprintf(format+"%s%s", value, space, units[pow])
}

func main() {
    if len(os.Args) < 2 {
        fmt.Println("usage: format_file_size <bytes> [precision]")
        os.Exit(1)
    }
    b, _ := strconv.ParseFloat(os.Args[1], 64)
    p := 2
    if len(os.Args) > 2 {
        if pp, err := strconv.Atoi(os.Args[2]); err == nil {
            p = pp
        }
    }
    fmt.Println(formatFileSize(b, p, true))
}


